### DISCUSSION TITLE
[Help wanted] Renovate unit tests against deprecated `findDOMNode` API

### DISCUSSION THREAD

**Discussion by SevenOutman** [rsuite/rsuite#3200] [Announcement]
Hi community, here's the first "good first issue" in rsuite repo (not an actual issue though), which means here're some tasks that we want to invite you to participate.

I think it's a good opportunity for you if you either:

- want to know better about rsuite
- want to learn how to unit-test React components
- want to get involved in some open-source projects and community

## What's the task?

TL;DR To update the way we used to render React components in test scripts - which is using `getDOMNode()` util from `test/testUtils.js` - to using React Testing Library `render()`.

### Background

Old unit tests in rsuite repo uses a util function called `getDOMNode()` to render a React component and retrieves its root element. This util function does not have proper cleanup and uses [the deprecated `findDOMNode` API](https://beta.reactjs.org/reference/react-dom/findDOMNode) under the hood. Last year we adopted React Testing Library for new tests, but left old tests as is. But as we're updating the React version, it becomes necessary to drop the usage of deprecated APIs.

## What could I learn?

- You will have better understanding on how rsuite components are implemented, tested and organized
- You will learn how to unit-test React components with React Testing Library
- You will get involved in the PR-review workflow of an active open-source project, and get code review from rsuite team members

## How do I contribute?

If you find yourself interested in contributing, please leave a comment below and tell me which components/modules you would like to work on. 

Here's a (long) list of components/modules that are using `getDOMNode()` util in their unit tests and needs to be renovated.

<details>
<summary>Expand</summary>

- Col (✅ done by @WWK563388548)
- Container (✅ done by @WWK563388548)
- Content (✅ done by @WWK563388548)
- CustomProvider (✅ done by @WWK563388548)
- DatePicker (✅ done by @WWK563388548)
- DateRangePicker (✅ done by @restartgr) 
- Divider (✅ done by @restartgr)
- Drawer (✅ done by @restartgr)
- Dropdown (✅ done by @restartgr)
- FlexboxGrid (✅ done by @restartgr)
- Footer (✅ done by @restartgr)
- Form (✅ done by @restartgr)
- FormControl (✅ done by @Chriscaracach)
- FormControlLabel (assigned to @restartgr)
- FormErrorMessage (assigned to @restartgr)
- FormGroup (assigned to @restartgr)
- FormHelpText (assigned to @restartgr)
- Grid (assigned to @restartgr)
- Header (assigned to @mx-in)
- IconButton (✅ done by @Chriscaracach)
- Input (assigned to @mx-in)
- InputGroup (assigned to @mx-in)
- InputNumber (✅ done by @Chriscaracach)
- InputPicker (assigned to @mx-in)
- List (assigned to @mx-in)
- Loader (assigned to @mx-in)
- Message (assigned to @mx-in)
- Modal (assigned to @mx-in)
- MultiCascader (assigned to @WWK563388548)
- Nav (assigned to @WWK563388548)
- Navbar (assigned to @WWK563388548)
- Overlay (assigned to @WWK563388548)
- Pagination (assigned to @WWK563388548)
- Panel (assigned to @WWK563388548)
- PanelGroup (assigned to @WWK563388548)
- Picker (assigned to @WWK563388548)
- Placeholder (assigned to @WWK563388548)
- Plaintext (assigned to @WWK563388548)
- Popover (assigned to @WWK563388548)
- Progress (assigned to @WWK563388548)
- Radio (assigned to @WWK563388548)
- RadioGroup (assigned to @WWK563388548)
- RangeSlider (assigned to @WWK563388548)
- RateRipple (assigned to @WWK563388548)
- Row (assigned to @WWK563388548)
- SafeAnchor (✅ done by @Gemini-Kang)
- SelectPicker (assigned to @Gemini-Kang)
- Sidebar (assigned to @Gemini-Kang)
- Sidenav (assigned to @Gemini-Kang)
- Slider (assigned to @Gemini-Kang)
- Steps (assigned to @Gemini-Kang)
- Tag (✅ done by @leon-fong)
- TagGroup (✅ done by @leon-fong)
- TagPicker (assigned to @Gemini-Kang)
- Timeline (assigned to @Gemini-Kang)
- toaster (assigned to @Gemini-Kang)
- Toggle (assigned to @Gemini-Kang)
- Tooltip (assigned to @Gemini-Kang)
- Tree (✅ done by @Chriscaracach)
- TreePicker (✅ done by @Chriscaracach)
- Uploader (✅ done by @Chriscaracach)
- utils (assigned to @Gemini-Kang)
- Whisper (✅ done by @Chriscaracach)

</details>

**You would checkout a new branch based on latest `main` branch, finish your edit, then open a pull request targeting `main` branch.**

For your reference, I'll post an additional comment to provide several examples of how to refactor the tests in specific scenarios.

Additionally, I've already updated some modules (excluded from the list) in https://github.com/rsuite/rsuite/pull/3046. You could always refer to that PR to see real-world examples.

**Comment by SevenOutman** [comment]
## The additional comment

This comment is meant to provide hints about how to refactor the tests in specific scenarios.

### Rendering a React component

To render a React component, old tests use `getDOMNode()` function, which does not have proper cleanup

```jsx
getDOMNode(<Component />);
```

New tests uses [`render()` function from React Testing Library](https://testing-library.com/docs/react-testing-library/api#render), which by default cleans-up after each test case

```jsx
render(<Component />);
```

### Rendering multiple instances of a React component with different props 

In old tests, we do this by calling `getDOMNode()` multiple times.

```jsx
const instance1 = getDOMNode(<Component {...props1} />);
const instance2 = getDOMNode(<Component {...props2} />);

expect(instance1).some.thing;
expect(instance2).some.thing;
```

In new tests, we use [`rerender` function](https://testing-library.com/docs/react-testing-library/api#rerender) returned by `render()` function

```jsx
const { rerender } = render(<Component {...props1} />);

expect(...).some.thing;

rerender(<Component {...props2} />);

expect(...).some.thing;
```

### Getting the root element of rendered component

To access the root element of rendered element, old tests uses the returned value of `getDOMNode()` function - which is what we are eliminating in this task.

```jsx
const instance = getDOMNode(<Component />);
```

In new tests, most of the time we don't directly access the root element, but follow [the guiding principles of React Testing Library](https://testing-library.com/docs/guiding-principles). But if it does consider the element detail (e.g. testing it has expected class names), most of time you can access the root element via the `ref` prop.

```jsx
const elementRef = render(<Component ref={elementRef} />);

expect(elementRef.current).some.thing;
```

For those components where `ref` doesn't return the root element, you could use `container.firstChild` returned by `render()`

```jsx
const { container } = render(<Component />)

expect(container.firstChild).some.thing;
``` 

### Testing class names, styles, attributes

To check that an element has expected class names, old tests cases might go

```jsx
const instance = getDOMNode(<Component />);

expect(instance.className).to.include('some-class');
// or
assert.include(instance.className, 'some-class');
```

New tests make use of the `.to.have.class()` assertion from [chai-dom](https://github.com/nathanboktae/chai-dom) library.

```jsx
const { container } = render(<Component />);

expect(container.firstChild).to.have.class('some-class');
```

Same rule applies to checking element's style and attributes. Use `.to.have.style()` and `.to.have.attr()` assertions from `chai-dom`.

### Common test cases

Most components has these 3 common test cases:
- Check that custom class names could be applied via `className` prop
- Check that custom style could be applied via `style` prop
- Check that custom class prefixes could be applied via `classPrefix` prop

In old tests, they usually look like this:

```jsx
it('Should have a custom className', () => {
  const instance = getDOMNode(<Component className="custom" />);
  assert.include(instance.className, 'custom');
});

it('Should have a custom style', () => {
  const fontSize = '12px';
  const instance = getDOMNode(<Component style={{ fontSize }} />);
  assert.equal(instance.style.fontSize, fontSize);
});

it('Should have a custom className prefix', () => {
  const instance = getDOMNode(<Component classPrefix="custom-prefix" />);
  assert.include(instance.className, 'custom-prefix');
});
```

In new tests, they are replaced by a test util `testStandardProps` (exported from `test/commonCases.js` which includes all 3 test cases.

```jsx
testStandardProps(<Component />)
```

**Comment by WWK563388548** [comment]
@SevenOutman 
Hi, Doma！Thank you very much for this very good opportunity.
I would like to pick up the first to fifth components to get a good start.
（Col，Container，Content，CustomProvider，DatePicker）

**Reply by SevenOutman** [reply]
Thanks for taking part and thanks for being the first to volunteer! I've marked these components you pick as assigned. Feel free to ask anything about the development while you roll!

**Comment by restartgr** [comment]
@SevenOutman 
Hello! I want to help with this issue, I want to select these components  (DateRangePicker,Divider,Drawer,Dropdown,FlexboxGrid,Footer,Form)

**Reply by SevenOutman** [reply]
Thanks for taking part, that's quite many components! I've marked these components you pick as assigned. Feel free to ask anything about the development while you roll!

**Reply by restartgr** [reply]
@SevenOutman 
Hi,I think i have free time to deal with that job, so i still want to select these componets

**Reply by SevenOutman** [reply]
I guess you mean FormControl,FormControlLabel,FormErrorMessage,FormGroup,FormHelpText,Grid as you mentioned in [this comment](https://github.com/rsuite/rsuite/discussions/3200#discussioncomment-5963857). I've marked these components as assigned to you. Enjoy your second roll!

**Comment by mx-in** [comment]
hi @SevenOutman, I'd like to help with components below (from Header -to Modal) during this weekend :)
- Header
- IconButton
- Input
- InputGroup
- InputNumber
- InputPicker
- List
- Loader
- Message
- Modal

**Reply by SevenOutman** [reply]
Thanks for taking part! I've marked these components you pick as assigned. Enjoy your coding weekend! Feel free to ask anything about the development while you roll!

**Comment by restartgr** [comment]
@SevenOutman 
Hello, I have completed my  previous tasks,and then I want to continue to apply for the task of modifying these components
(FormControl,FormControlLabel,FormErrorMessage,FormGroup,FormHelpText,Grid)

**Reply by SevenOutman** [reply]
Wow, that's so quick! I would suggest we get you first PR reviewed and merged first, but it's alright if you want to pick more tasks right now :). Please confirm in a reply if you decide to take new tasks now, and I'll assign those modules to you.

**Reply by restartgr** [reply]
I think that I should deal with my first PR first.After I merged the PR,I will apply to pick more component

**Comment by leon-fong** [comment]
@SevenOutman 
Hi, I'm interested in the Tag component task as I'm currently learning to write unit tests for React components using React Testing Library.

**Reply by SevenOutman** [reply]
Thanks for taking part! I've marked Tag as assigned to you. Have a good time testing! Feel free to ask anything about the development while you roll!

BTW, would you like to also take TagGroup, as Tag and TagGroup are quite closely related?

**Reply by leon-fong** [reply]
Yes, that's a great idea! I would love to include TagGroup as well. Thank you for suggesting it.

**Reply by SevenOutman** [reply]
Great! I've marked TagGroup as assigned to you as well.

**Comment by WWK563388548** [comment]
@SevenOutman 
Hello! I have completed the task that I previously claimed and would like to continue receiving more tasks: 
- MultiCascader
- Nav
- Navbar
- Overlay
- Magination
- Panel
- PanelGroup
- Picker
- Placeholder
- Plaintext
- Popover
- Progress
- Radio
- RadioGroup
- RangeSlider
- RateRipple
- Row

By the way, if you have the time, I would appreciate it if you could review these pull requests related to the previous task:
https://github.com/rsuite/rsuite/pull/3262 (CustomProvider & Content)
https://github.com/rsuite/rsuite/pull/3261 (DatePicker)
https://github.com/rsuite/rsuite/pull/3258 (Container)
https://github.com/rsuite/rsuite/pull/3257 (Col)

**Reply by SevenOutman** [reply]
Thanks for being so helpful! I've peeked those PR and they generally look good to me. I've marked these components as assigned to you, happy coding & testing!

**Comment by Gemini-Kang** [comment]
@SevenOutman 
Hello!
Thank you very much for this very good opportunity.
I would like to pick up some components to get a good start.
（SafeAnchor, SelectPicker, Sidebar, Sidenav, Slider, Steps, TagPicker, Timeline, toaster, Toggle, Tooltip）

**Reply by SevenOutman** [reply]
Thanks for taking part! I've marked these components you pick as assigned to you. Hope you enjoy the coding & testing! Feel free to ask anything about the development while you roll!

**Comment by Chriscaracach** [comment]
@SevenOutman 
Hi!
I'd also like to help with this. 
I'm picking this components:
(Whisper, Uploader, Treepicker)

**Reply by SevenOutman** [reply]
Thanks for being willing to help! I've marked these components you pick as assigned to you. Enjoy the game and feel free to ask anything about the development while you roll!

BTW, would you like to also take Tree, as Tree and TreePicker are quite closely related?

**Reply by Chriscaracach** [reply]
Yes. No problem, I didn't saw that it was available. I'll take care of it.

**Reply by SevenOutman** [reply]
That's great! I've marked Tree as assigned to you as well.

**Comment by WWK563388548** [comment]
@SevenOutman 
There is one `utils` left. Does `utils` mean a series of tests in `rsuite/src/utils`?
If so, you can assign it to @Gemini-Kang . She thinks she can fix it

**Reply by SevenOutman** [reply]
> @SevenOutman There is one `utils` left. Does `utils` mean a series of tests in `rsuite/src/utils`? 

Yes `utils` refers to tests in `src/utils` directory.

> If so, you can assign it to @Gemini-Kang . She thinks she can fix it

@Gemini-Kang Could you please confirm this?

**Reply by Gemini-Kang** [reply]
@SevenOutman 
That's true.
Please assign it to me.
let's close this discussion~

**Reply by SevenOutman** [reply]
That's great. I've marked `utils` as assigned to you as well.

**Comment by Chriscaracach** [comment]
Question over here. What do you prefer, this way:
`// eslint-disable-next-line testing-library/no-node-access, testing-library/no-container`
or this way:
`// eslint-disable-next-line`
?

**Reply by SevenOutman** [reply]
I prefer explicitly declaring the rules to disable, so that others (or me afterwards) get more info about what's going on here.

**Reply by Chriscaracach** [reply]
Hi, sorry for the delay. Wild weeks at work. UploaderTrigger isn't passing "should accept custom style" from testStandardProps(). Only that one. I'm missing something here? I saw this small difference on the older way to test it. Maybe it's related:
```
it('Should have a custom style', () => {
    const fontSize = '12px';
    const instance = getDOMNode(<UploadTrigger style={{ fontSize }} />);
    assert.equal((instance.querySelector('button') as HTMLElement).style.fontSize, fontSize); //This line
  });
```
Any idea?

**Reply by SevenOutman** [reply]
This is because `testStandardProps()` checks `style` property on the root element that the component renders, but UploaderTrigger does not seem to apply `style` on its root element. Setting `getRootElement: view => view.getByRole('button')` would make the style test pass (like [what we do with SidenavToggle](https://github.com/rsuite/rsuite/pull/3046/files#diff-1eaa9e7cf82bbaf6eed0ae3121a0c7a4dabf7e38695e3a2898394fbcaaee3689), but that would fail the className test, and that's because UploaderTrigger applies `className` on the root element. I would suggest you leave the related cases as is.

**Comment by Nepsha1986** [comment]
Hi community! I've tried to make a few tests but got some strange behavior. 
Let's say I want to change a test for the ModalDialog component and I'm running:
```M=ModalDialog npm run tdd```
Then I made some changes, but I do not see the updated test run. And even if I manually rerun the test from the IDEA ( I use WebStorm ) I do not see changes has been applied, so I  I need to stop the test runner ( karma ) and again run "M=ModalDialog npm run tdd" to see if the test passes.  Does anybody of you have the same behavior?

**Reply by simonguo** [reply]
The `M` environment variable actually specifies the directory of a component. ModalDialog is not a directory, so the test cannot be run. 

https://github.com/rsuite/rsuite/blob/main/karma.conf.js#L21

```bash
# only run tests in src/Modal directory
M=Modal npm run tdd
```

```bash
# only run tests in src/Modal/test/ModalDialogSpec.tsx
F=src/Modal/test/ModalDialogSpec.tsx npm run tdd
```

**Reply by Nepsha1986** [reply]
Ok, I get it. Thank you for explanation @simonguo )

**Comment by Chriscaracach** [comment]
Hello there! Is there any update on this? I'm still willing to help if u need. Can I take some of the components that doesn't have any progress yet?

**Reply by simonguo** [reply]
Hi @Chriscaracach , we still need help, thank you for your support.

**Reply by Chriscaracach** [reply]
Ok! @WWK563388548 Are you working on the ones that has been assigned to you? Please answer me, if not, I can take care of them.

**Reply by Chriscaracach** [reply]
Updates here! I've created a PR with changes for Navbar, PanelGroup and Popover. 
You may want also mark as "done" the following components:
- Multicascader
- Nav
- Pagination
- Panel

Those were already refactored, some of them by @WWK563388548 , others by @simonguo , I think. (not sure)

**Reply by simonguo** [reply]
Yes, there are some components that I refactored the test cases when fixing bugs

**Comment by Chriscaracach** [comment]
@mx-in @restartgr Are you working on this? I can take care of the ones assigned to you. Please let me know. No pressure. =)

**Comment by Chriscaracach** [comment]
Hi! I think I did the last PR related to this discussion. Please take a look @simonguo.
There are some components that still use getDOMNode, but they're all under src/internals folder. And I think that's outside the scope of this issue. 
So, that's it. Please let me know if there is something else to do about this, I'm still willing to help. If not, I think this discussion is done.

**Reply by simonguo** [reply]
Hi  @Chriscaracach  
Thank you for your contribution, I will merge this PR as soon as possible. If you are interested, you can continue to help fix other places (including under the src/internals directory), we will be very grateful.
